supervisor/qemu/src/main.rs

```diff
@@ -1180,44 +1198,77 @@ impl connector::Supervisor for QemuSupervisor {
 
         // Perform actions depending on the previous job state:
         match prev_job_state {
-            StopJobPrevState::FetchingImage(QemuSupervisorFetchingImageState {
-                start_job_req: _,
-		image_id: _,
-                poll_task,
-            }) => {
-                // Make sure that we don't have another poll task
-                // scheduled. This is potentially racy, where this abort() call
-                // can come right after an invocation of `Self::fetch_image` has
-                // been scheduled. However, that function will not do anything,
-                // given the job state is now `Stopping` instead of
-                // `FetchingImage`:
-                if let Some(handle) = poll_task {
+            QemuSupervisorJobState::FetchingImage(state) => {
+                if let Some(handle) = state.poll_task {
                     handle.abort();
                 }
             }
-
-            StopJobPrevState::ImageFetched(QemuSupervisorImageFetchedState {
-                start_job_req: _,
-		image_id: _,
-                manifest: _,
-            }) => {
+            QemuSupervisorJobState::ImageFetched(_) => {
                 // Nothing to do, the only time we're seeing this state is in
                 // between fetching image, and actually starting the job. Given
                 // that we've acquired the lock between those two functions, we
                 // don't need to deallocate, shut down or abort anything.
             }
-
-            StopJobPrevState::Running(QemuSupervisorJobRunningState {
-                start_job_req: _,
-                control_socket,
-                mut qemu_proc,
-            }) => {
+            QemuSupervisorJobState::Running(mut running_state) => {
                 // TODO: kindly request the puppet to shut down. Here we simply
                 // force it to quit (by using a SIGKILL). This is not nice.
-                qemu_proc.kill().await.unwrap();
+                let shutdown_result = async {
```

Comment:
We hold the job lock over the entire duration of this async block (currently up to 5 minutes). I think it might be better if we request the shutdown monitor command and then set the job state to stopping. After 5 minutes, we can invoke another function that then proceeds to `SIGTERM` the QEMU process if it didn't already shut down. This would use a very similar logic to the image fetching state, where it uses async recursion to call itself after some time has elapsed -- without holding the lock during that wait period.